### ISSUE TITLE
training problem with imbalanced data

### ISSUE THREAD

**Issue by DF-erlian** [PerfVec/PerfVec#2]
Hi,
   Have you ever faced a training problem with imbalanced data? For example, when the labels for fetch latency are mostly 0 or 1, and the range of labels spans from 0 to over 1000. Some labels appear frequently, while most others are very rare. Additionally, the distribution can vary significantly across different workloads or even within different parts of the same workload.
   When I try to train a model, it often fails to converge—the loss value stabilizes at a certain point and doesn't decrease further. I've tried several tricks, such as label smoothing, loss re-weighting, and combining regression with classification, but none of them have worked.
   I'm not very experienced in model training and don't know where to go. Could you give me some suggestions?
Thank you!

**Comment by lingda-li** [comment]
Thanks for your interest to our work.

Imbalance distribution of latencies should be common in the training data. It didn't necessarily create a problem in our training. Nevertheless it could potentially result in better models if you manage to improve it.

About loss convergence, there are many possible causes. We observed loss stablizes too and it does not necessarily mean the trained model is not accuarte if you test it.

**Comment by DF-erlian** [comment]
Thanks for your reply!
I successfully reproduced your code by following your tutorial.

I tried training "InsLSTM(256, 2, narchs=1, bias=False)" from scratch using training data selected from the `gemm60x70x80` program, which contains 1,551,224 execution instructions.

Config file:
```
input_length = 51
tgt_length = 1
cfg_num = 1
seq_length = 256
ori_tgt_length = 16

def sel_batch_out(y):
  y = y.reshape((-1, cfg_num, ori_tgt_length))
  y = y[:, :, 0] # fetch latency
  return y

``` 
Training command:
```
python -m ML.train --cfg=com_shixl --train-size=1550848 --epochs=200 --batch-size=2048 --sbatch "InsLSTM(256,2,narchs=1,bias=False)" --lr=0.001 --lr-step=50
```
The model get best checkpoint at epoch 151 with Training Loss 257.902353 Test Loss 167.702467
Sim result:
```shell
...
Test loss: 268.711888   Time: 47.5 s
(1536000, 1, 1)
Config 0
0 :
        target: [800.   5.   0. ...   5.   0.   5.]
        norm output: [802.   2.   0. ...   5.  -0.   4.]
        errors: [-2.  3.  0. ...  0.  0.  1.]
        Error abs avg, norm abs avg, norm avg, and norm std: 4.9091945   1.8210762       -1.4744631      5.291573
...
Run 1551224 instructions.
trace
Use the maximum size of 3029 instead of 3030
Target: tensor([[11909115.]])
Output: tensor([[12009564.]])
Error: tensor([[0.0084]])
Mean error: tensor([0.0084])
Mean normalized error: tensor([0.0084])
Loss: 0.088510  Time: 41.0
```
The result is remarkable! However, I still have some confusion:
1. Is the training loss acceptable, or could the issue be caused by the dataset size or insufficient epochs?
2. I got an absolute average error value of 4.909, but when summing the latency of each instruction, the error decreased to 0.0084 per instruction. This indicates that the errors for individual instructions may have both positive and negative values, which mostly cancel out during the summation process. Is this a coincidence, or is it an inherent effect of the model?

**Comment by lingda-li** [comment]
For your questions:

1. It's hard to tell what loss is acceptable or not as it depends on many factors (e.g., batch size, dataset).
2. These errors have different units. 4.909 is for instruction latency, and 0.0084 is the percentage error for the whole program. So they are not directly comparable. We do expect instruction errors to cancel out each other at the program level to reach better accuracy.

**Comment by liuqinfei** [comment]
@DF-erlian I am instrested in the performance modeling and predicting. It semms that you already run the project well, Do you know how to tune the configurations of arm platform when doing performance predicting?
The project is so nice. If you are free, please reply. Thanks. @lingda-li

**Comment by DF-erlian** [comment]
I reproduced the code by following the markdown files under `Tutorials`, such as `gem5.md` for generating input traces, `learn_rep.md` for obtaining instruction representations, `predict.md` for generating simulation results, and `train.md` for training a model from scratch.

**Comment by liuqinfei** [comment]
> I reproduced the code by following the markdown files under `Tutorials`, such as `gem5.md` for generating input traces, `learn_rep.md` for obtaining instruction representations, `predict.md` for generating simulation results, and `train.md` for training a model from scratch.

Yes, i also reprodeced the steps successfully except train.d as my GPU server is not ready yet. But, did you know how to tune parameters of platform such as the size of L3 cache?

**Comment by DF-erlian** [comment]
Chapter VI of the author's paper "Learning Generalizable Program and Architecture Representations for Performance Modeling" may be helpful, which introduces the applications of PrefVec.

**Comment by lingda-li** [comment]
@liuqinfei You can change the configuration in gem5. PerfVec requires a learned microarchitecture representation in that case to make prediciton.
